### ISSUE TITLE
Incorrect outdoor temp

### ISSUE THREAD

**Issue by AngryBalloon** [toringer/home-assistant-heru#17]
## Version of the custom_component

Version 1.0.1

## Configuration
Heru 100 S EC y1
Home Assistant 2023.10.3

## Describe the bug
Outdoor temperature just dropped below 0°C and now Outdoor temperature shows a value of 6553.2°C in HA.
Wireless control unit shows 0°C so this should not be a sensor issue.

Graph in HA shows temperature steadily decresing with last value of 0.2°C before hitting 6553.5°C

Other temps show correct value. Outdoor temp has been correct until now. There has only been temps above 0°C as long as I have used this component.

Edit: As soon as the outdoor temperature rised above 0°C again value presented in HA returned to showing correct value.

## Debug log

```text

2023-10-19 20:38:07.895 DEBUG (MainThread) [pymodbus.logging] send: 0x0 0x31 0x0 0x0 0x0 0x6 0x1 0x1 0x0 0x0 0x0 0x7
2023-10-19 20:38:07.897 DEBUG (MainThread) [pymodbus.logging] Adding transaction 49
2023-10-19 20:38:08.024 DEBUG (MainThread) [pymodbus.logging] recv: 0x0 0x31 0x0 0x0 0x0 0x4 0x1 0x1 0x1 0x1
2023-10-19 20:38:08.025 DEBUG (MainThread) [pymodbus.logging] Processing: 0x0 0x31 0x0 0x0 0x0 0x4 0x1 0x1 0x1 0x1
2023-10-19 20:38:08.025 DEBUG (MainThread) [pymodbus.logging] Factory Response[ReadCoilsResponse': 1]
2023-10-19 20:38:08.026 DEBUG (MainThread) [pymodbus.logging] Getting transaction 49
2023-10-19 20:38:08.027 DEBUG (MainThread) [pymodbus.logging] send: 0x0 0x32 0x0 0x0 0x0 0x6 0x1 0x2 0x0 0x0 0x0 0x36
2023-10-19 20:38:08.029 DEBUG (MainThread) [pymodbus.logging] Adding transaction 50
2023-10-19 20:38:08.055 DEBUG (MainThread) [pymodbus.logging] recv: 0x0 0x32 0x0 0x0 0x0 0xa 0x1 0x2 0x7 0x0 0x0 0x0 0x40 0x0 0x0 0x0
2023-10-19 20:38:08.056 DEBUG (MainThread) [pymodbus.logging] recv: 0x0 0x32 0x0 0x0 0x0 0xa 0x1 0x2 0x7 0x0 0x0 0x0 0x40 0x0 0x0 0x0
2023-10-19 20:38:08.057 DEBUG (MainThread) [pymodbus.logging] Processing: 0x0 0x32 0x0 0x0 0x0 0xa 0x1 0x2 0x7 0x0 0x0 0x0 0x40 0x0 0x0 0x0
2023-10-19 20:38:08.057 DEBUG (MainThread) [pymodbus.logging] Factory Response[ReadDiscreteInputsResponse': 2]
2023-10-19 20:38:08.058 DEBUG (MainThread) [pymodbus.logging] Getting transaction 50
2023-10-19 20:38:08.065 DEBUG (MainThread) [pymodbus.logging] send: 0x0 0x33 0x0 0x0 0x0 0x6 0x1 0x4 0x0 0x0 0x0 0x21
2023-10-19 20:38:08.066 DEBUG (MainThread) [pymodbus.logging] Adding transaction 51
2023-10-19 20:38:08.076 DEBUG (MainThread) [pymodbus.logging] recv: 0x0 0x33 0x0 0x0 0x0 0x45 0x1 0x4 0x42 0x0 0xa 0xff 0xfc 0x0 0x96 0x0 0xb6 0x0 0x2d 0xd8 0xfa 0x0 0x99 0xd8 0xfa 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x50 0x0 0x0 0x1 0x2f 0x0 0x0 0x0 0x0 0x0 0x2 0x0 0x2 0x0 0x3c 0x0 0x41 0x8 0x9c 0x9 0x9f 0x0 0x0 0x0 0xff 0x0 0x0 0x0 0x3c 0x0 0x41
2023-10-19 20:38:08.077 DEBUG (MainThread) [pymodbus.logging] recv: 0x0 0x33 0x0 0x0 0x0 0x45 0x1 0x4 0x42 0x0 0xa 0xff 0xfc 0x0 0x96 0x0 0xb6 0x0 0x2d 0xd8 0xfa 0x0 0x99 0xd8 0xfa 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x50 0x0 0x0 0x1 0x2f 0x0 0x0 0x0 0x0 0x0 0x2 0x0 0x2 0x0 0x3c 0x0 0x41 0x8 0x9c 0x9 0x9f 0x0 0x0 0x0 0xff 0x0 0x0 0x0 0x3c 0x0 0x41
2023-10-19 20:38:08.077 DEBUG (MainThread) [pymodbus.logging] Processing: 0x0 0x33 0x0 0x0 0x0 0x45 0x1 0x4 0x42 0x0 0xa 0xff 0xfc 0x0 0x96 0x0 0xb6 0x0 0x2d 0xd8 0xfa 0x0 0x99 0xd8 0xfa 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x50 0x0 0x0 0x1 0x2f 0x0 0x0 0x0 0x0 0x0 0x2 0x0 0x2 0x0 0x3c 0x0 0x41 0x8 0x9c 0x9 0x9f 0x0 0x0 0x0 0xff 0x0 0x0 0x0 0x3c 0x0 0x41
2023-10-19 20:38:08.078 DEBUG (MainThread) [pymodbus.logging] Factory Response[ReadInputRegistersResponse': 4]
2023-10-19 20:38:08.078 DEBUG (MainThread) [pymodbus.logging] Getting transaction 51
2023-10-19 20:38:08.079 DEBUG (MainThread) [pymodbus.logging] send: 0x0 0x34 0x0 0x0 0x0 0x6 0x1 0x3 0x0 0x0 0x0 0x43
2023-10-19 20:38:08.080 DEBUG (MainThread) [pymodbus.logging] Adding transaction 52
2023-10-19 20:38:08.092 DEBUG (MainThread) [pymodbus.logging] recv: 0x0 0x34 0x0 0x0 0x0 0x89 0x1 0x3 0x86 0x0 0xf 0x0 0x15 0x0 0x3c 0x0 0x41 0x0 0x14 0x0 0x50 0x0 0x96 0x0 0xa0 0x0 0x3c 0x1 0x4 0x0 0x0 0x0 0x0 0x0 0xf 0x0 0x23 0x0 0x2 0x0 0xc 0x0 0x5 0x0 0x0 0x0 0x0 0x0 0x32 0x0 0x19 0x0 0x14 0x0 0x0 0x0 0x4 0x0 0xf 0x0 0x1e 0x0 0x1e 0x0 0xf 0x0 0x19 0x0 0x0 0x0 0x0 0x0 0x50 0x0 0x50 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x1 0x0 0x0 0x0 0x50 0x0 0x50 0x0 0xc 0xff 0x68 0xff 0xff 0x0 0x0 0x0 0x28 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0xe 0x0 0x10 0x0 0x18 0x4 0xb1 0x2 0x59 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x1e 0x0 0x0 0x0 0x0 0xff 0xf1 0x0 0x2 0x0 0x0
2023-10-19 20:38:08.093 DEBUG (MainThread) [pymodbus.logging] recv: 0x0 0x34 0x0 0x0 0x0 0x89 0x1 0x3 0x86 0x0 0xf 0x0 0x15 0x0 0x3c 0x0 0x41 0x0 0x14 0x0 0x50 0x0 0x96 0x0 0xa0 0x0 0x3c 0x1 0x4 0x0 0x0 0x0 0x0 0x0 0xf 0x0 0x23 0x0 0x2 0x0 0xc 0x0 0x5 0x0 0x0 0x0 0x0 0x0 0x32 0x0 0x19 0x0 0x14 0x0 0x0 0x0 0x4 0x0 0xf 0x0 0x1e 0x0 0x1e 0x0 0xf 0x0 0x19 0x0 0x0 0x0 0x0 0x0 0x50 0x0 0x50 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x1 0x0 0x0 0x0 0x50 0x0 0x50 0x0 0xc 0xff 0x68 0xff 0xff 0x0 0x0 0x0 0x28 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0xe 0x0 0x10 0x0 0x18 0x4 0xb1 0x2 0x59 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x1e 0x0 0x0 0x0 0x0 0xff 0xf1 0x0 0x2 0x0 0x0
2023-10-19 20:38:08.093 DEBUG (MainThread) [pymodbus.logging] Processing: 0x0 0x34 0x0 0x0 0x0 0x89 0x1 0x3 0x86 0x0 0xf 0x0 0x15 0x0 0x3c 0x0 0x41 0x0 0x14 0x0 0x50 0x0 0x96 0x0 0xa0 0x0 0x3c 0x1 0x4 0x0 0x0 0x0 0x0 0x0 0xf 0x0 0x23 0x0 0x2 0x0 0xc 0x0 0x5 0x0 0x0 0x0 0x0 0x0 0x32 0x0 0x19 0x0 0x14 0x0 0x0 0x0 0x4 0x0 0xf 0x0 0x1e 0x0 0x1e 0x0 0xf 0x0 0x19 0x0 0x0 0x0 0x0 0x0 0x50 0x0 0x50 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x1 0x0 0x0 0x0 0x50 0x0 0x50 0x0 0xc 0xff 0x68 0xff 0xff 0x0 0x0 0x0 0x28 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0xe 0x0 0x10 0x0 0x18 0x4 0xb1 0x2 0x59 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x1e 0x0 0x0 0x0 0x0 0xff 0xf1 0x0 0x2 0x0 0x0
2023-10-19 20:38:08.094 DEBUG (MainThread) [pymodbus.logging] Factory Response[ReadHoldingRegistersResponse': 3]
2023-10-19 20:38:08.094 DEBUG (MainThread) [pymodbus.logging] Getting transaction 52
2023-10-19 20:38:08.095 DEBUG (MainThread) [custom_components.heru.heru_coordinator] Finished fetching HERU data in 0.201 seconds (success: True)
2023-10-19 20:38:08.095 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.096 DEBUG (MainThread) [custom_components.heru.sensor] Outdoor temperature: 6553.200000000001 °C
2023-10-19 20:38:08.096 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.096 DEBUG (MainThread) [custom_components.heru.sensor] Supply air temperature: 15.0 °C
2023-10-19 20:38:08.097 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.097 DEBUG (MainThread) [custom_components.heru.sensor] Extract air temperature: 18.2 °C
2023-10-19 20:38:08.098 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.098 DEBUG (MainThread) [custom_components.heru.sensor] Exhaust air temperature: 4.5 °C
2023-10-19 20:38:08.098 DEBUG (MainThread) [custom_components.heru.sensor] Heat recovery temperature: 15.3 °C
2023-10-19 20:38:08.099 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.099 DEBUG (MainThread) [custom_components.heru.sensor] Current exhaust fan power: 65 %
2023-10-19 20:38:08.100 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.100 DEBUG (MainThread) [custom_components.heru.sensor] Current heating power: 0.0 %
2023-10-19 20:38:08.101 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.102 DEBUG (MainThread) [custom_components.heru.sensor] Current supply fan power: 60 %
2023-10-19 20:38:08.102 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.102 DEBUG (MainThread) [custom_components.heru.sensor] Current heat/cold recovery power: 99.9999999885 %
2023-10-19 20:38:08.103 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.103 DEBUG (MainThread) [custom_components.heru.sensor] Night cooling active: off None
2023-10-19 20:38:08.104 DEBUG (MainThread) [custom_components.heru.sensor] Filter timer alarm: off None
2023-10-19 20:38:08.104 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.104 DEBUG (MainThread) [custom_components.heru.sensor] Exhaust fan alarm: off None
2023-10-19 20:38:08.105 DEBUG (MainThread) [custom_components.heru.sensor] Boost input: off None
2023-10-19 20:38:08.105 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.105 DEBUG (MainThread) [custom_components.heru.sensor] Overpressure input: off None
2023-10-19 20:38:08.107 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.108 DEBUG (MainThread) [custom_components.heru.sensor] Fire alarm: off None
2023-10-19 20:38:08.108 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.109 DEBUG (MainThread) [custom_components.heru.sensor] Rotor alarm: off None
2023-10-19 20:38:08.109 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.109 DEBUG (MainThread) [custom_components.heru.sensor] Supply fan alarm: off None
2023-10-19 20:38:08.110 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.110 DEBUG (MainThread) [custom_components.heru.sensor] Current supply fan step: Standard None
2023-10-19 20:38:08.110 DEBUG (MainThread) [custom_components.heru.sensor] Current exhaust fan step: Standard None
2023-10-19 20:38:08.111 DEBUG (MainThread) [custom_components.heru.sensor] Filter days left: 303 days
2023-10-19 20:38:08.111 DEBUG (MainThread) [custom_components.heru.sensor] HeruSensor._handle_coordinator_update()
2023-10-19 20:38:08.111 DEBUG (MainThread) [custom_components.heru.sensor] Startup 1st phase: off None
2023-10-19 20:38:08.111 DEBUG (MainThread) [custom_components.heru.sensor] Startup 2nd phase: off None
2023-10-19 20:38:08.112 DEBUG (MainThread) [custom_components.heru.switch] HeruSwitch._handle_coordinator_update()
2023-10-19 20:38:08.112 DEBUG (MainThread) [custom_components.heru.switch] Overpressure mode: False
2023-10-19 20:38:08.113 DEBUG (MainThread) [custom_components.heru.switch] HeruSwitch._handle_coordinator_update()
2023-10-19 20:38:08.113 DEBUG (MainThread) [custom_components.heru.switch] Boost mode: False
2023-10-19 20:38:08.113 DEBUG (MainThread) [custom_components.heru.switch] Away mode: False
2023-10-19 20:38:08.114 DEBUG (MainThread) [custom_components.heru.switch] Preheater enabled: 0
2023-10-19 20:38:08.114 DEBUG (MainThread) [custom_components.heru.switch] HeruSwitch._handle_coordinator_update()
2023-10-19 20:38:08.114 DEBUG (MainThread) [custom_components.heru.switch] Heater enabled: 0
2023-10-19 20:38:08.115 DEBUG (MainThread) [custom_components.heru.switch] Night cooling enabled: 0
2023-10-19 20:38:08.115 DEBUG (MainThread) [custom_components.heru.climate] HeruThermostat._handle_coordinator_update()
2023-10-19 20:38:08.116 DEBUG (MainThread) [custom_components.heru.climate] Comfort: 15.000000 21.000000
2023-10-19 20:38:08.116 DEBUG (MainThread) [custom_components.heru.number] HeruNumber._handle_coordinator_update()
2023-10-19 20:38:08.117 DEBUG (MainThread) [custom_components.heru.number] Night cooling indoor-outdoor diff. limit: 5.0 °C
2023-10-19 20:38:08.117 DEBUG (MainThread) [custom_components.heru.number] HeruNumber._handle_coordinator_update()
2023-10-19 20:38:08.118 DEBUG (MainThread) [custom_components.heru.number] Night cooling exhaust high limit: 25 °C
2023-10-19 20:38:08.118 DEBUG (MainThread) [custom_components.heru.number] HeruNumber._handle_coordinator_update()
2023-10-19 20:38:08.118 DEBUG (MainThread) [custom_components.heru.number] Night cooling exhaust low limit: 20 °C

```

**Comment by magicus** [comment]
Funny signed/unsigned bug :-)

**Comment by AngryBalloon** [comment]
I did a simple test and configured the [Modbus integration](https://www.home-assistant.io/integrations/modbus/) (which also uses pymodbus) to run in parallel and collect the same value. It does return the correct value for negative temperatures.



Configured it like this :
```
modbus:
  - name: ftx
    host: 192.168.1.14
    type: tcp
    port: 502
    
    sensors:
      - name: ftx_outdoor_temperature
        address: 1
        scale: 0.1
        unit_of_measurement: °C
        device_class: temperature
        input_type: input
```

Hope this can provide some help in the right direction.

**Comment by magicus** [comment]
I think the actual read happens in HeruCoordinator, on this line:
```
                result = await self.client.read_input_registers(0, 33, DEFAULT_SLAVE)
```

where client is a pymodbus client. I tried to read the pymodbus docs to see if this just defaulted to unsigned values, and/or if it was possible to change the type, but frankly I did not understand much about it.

If it is not possible to get pymodbus to deliver signed values, then I guess the integration needs to convert the values from unsigned to 2-complement's signed.

**Comment by magicus** [comment]
It seems complicated. I think this relates to the same problem we are facing: https://stackoverflow.com/questions/55662896/how-should-negative-numbers-be-represented-in-the-pymodbus-input-register

**Comment by kosmiq** [comment]
Just discovered the very same thing today!

@toringer can you add other maintainers that can approve PRs if you won’t work on the project? I love the implementation and would love to see merged PRs and fixes going forward :)

**Comment by magicus** [comment]
A workaround in the meantime is to add a template helper with this template:
```
{{ (states('sensor.ftx_outdoor_temperature') | float - 6553.5) | round(1) }}
```

Name the sensor e.g. `ftx_outdoor_temperature_real`. Set unit to °C and class to Temperature, and adjust the sensor name `ftx_outdoor_temperature` unless your heru is named `ftx`. Then use `ftx_outdoor_temperature_real` in place of `ftx_outdoor_temperature`.

**Comment by magicus** [comment]
Update: Use this template instead to handle both positive and negative temperatures :-)

```
{% set original_sensor = states('sensor.ftx_outdoor_temperature') | float %}
{% if original_sensor | float > 3200 %}
  {{ (original_sensor | float - 6553.5) | round(1) }}
{% else %}
  {{ original_sensor }}
{% endif %}
```

**Comment by toringer** [comment]
Sorry for the late respons. Added a temporary fix for the wrong value in v.1.0.2.

**Comment by lissmeister** [comment]
@toringer Thanks! The temperature in Stockholm just went from 6552,2°C to -1,3°C 🙂

**Comment by kosmiq** [comment]
Thanks! Works great here as well. Much appreciated :)

**Comment by AngryBalloon** [comment]
Thanks, can confirm it works for me as well!

**Comment by toringer** [comment]
Fixed permanently in v1.0.3
